plugins/microservices/src/json_value.cpp

```diff
@@ -0,0 +1,150 @@
+/// \file
+
+#include "msi_preconditions.hpp"
+
+#include "irods/irods_error.hpp"
+#include "irods/irods_logger.hpp"
+#include "irods/irods_ms_plugin.hpp"
+#include "irods/irods_re_structs.hpp"
+#include "irods/msParam.h"
+#include "irods/rodsErrorTable.h"
+#include "irods/storage.hpp"
+
+#include <boost/any.hpp>
+#include <nlohmann/json.hpp>
+
+#include <functional>
+#include <string>
+
+namespace
+{
+    // clang-format off
+    namespace is  = irods::experimental::storage;
+
+    using log_msi = irods::experimental::log::microservice;
+    using json    = nlohmann::json;
+    // clang-format on
+
+    // NOLINTNEXTLINE(bugprone-easily-swappable-parameters)
+    auto msi_impl(MsParam* _json_handle, MsParam* _json_pointer, MsParam* _json_value, ruleExecInfo_t* _rei) -> int
+    {
+        IRODS_MSI_REQUIRE_VALID_POINTER(_json_handle);
+        IRODS_MSI_REQUIRE_VALID_POINTER(_json_pointer);
+        IRODS_MSI_REQUIRE_VALID_POINTER(_json_value);
+        IRODS_MSI_REQUIRE_VALID_POINTER(_rei);
+
+        IRODS_MSI_REQUIRE_TYPE(_json_handle->type, STR_MS_T);
+        IRODS_MSI_REQUIRE_TYPE(_json_pointer->type, STR_MS_T);
+
+        IRODS_MSI_REQUIRE_VALID_POINTER(_json_handle->inOutStruct);
+        IRODS_MSI_REQUIRE_VALID_POINTER(_json_pointer->inOutStruct);
+
+        try {
+            auto* handle = static_cast<char*>(_json_handle->inOutStruct);
+            log_msi::debug("JSON handle (address=[{}], value=[{}])", static_cast<void*>(handle), handle);
+
+            auto* object = is::find(handle);
+            if (!object) { // NOLINT(readability-implicit-bool-conversion)
+                log_msi::error("Invalid argument: JSON handle [{}] is not associated with any object.", handle);
+                return SYS_INVALID_INPUT_PARAM;
+            }
+
+            const auto& json_object = boost::any_cast<json&>(*object);
+
+            const auto* p = static_cast<const char*>(_json_pointer->inOutStruct);
+            // NOLINTNEXTLINE(readability-implicit-bool-conversion)
+            log_msi::debug("JSON pointer address=[{}], value=[{}]", static_cast<const void*>(p), (p ? p : "<nullptr>"));
+
+            const json* json_value{};
+
+            try {
+                json_value = &json_object.at(json::json_pointer{p});
+            }
+            catch (const json::out_of_range&) {
+                fillStrInMsParam(_json_value, "<irods::json::out_of_range>");
+                return 0;
+            }
+
+            if (json_value->is_number()) {
+                const auto v = json_value->get<double>(); // Covers integers and floats.
+                fillStrInMsParam(_json_value, std::to_string(v).c_str());
+            }
```

Comment:
Should numbers be returned as numbers or strings?

Reply:
Yes. This enables policy writers to log the real value of a number. And the NREP supports converting strings to integers and floats.